<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <!-- <node pkg="crazyCmd" type="dd_optic_flow_data_converter.py" name="optic_flow_data_converter" /> -->
  <!-- <node pkg="crazyCmd" type="dd_optic_flow_data_converter_pose.py" name="optic_flow_data_converter_pose" /> -->
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization_custom" clear_params="true">
    <param name="frequency" value="30"/> <!-- this value should be fine -->
    <param name="sensor_timeout" value="0.1"/> <!-- this value should be fine -->
    <param name="two_d_mode" value="false"/> <!-- could make this true to stick to XY plane.. -->

    <!-- <param name="map_frame" value="map"/> -->
      <param name="map_frame" value="map"/>
      <param name="odom_frame" value="world"/>
      <param name="base_link_frame" value="crazyflie_main_body"/>
      <param name="world_frame" value="world"/>

    <param name="odom0" value="/optical_flow_twist"/>
    <!-- <param name="odom1" value="/optical_flow_pose"/> -->
    <param name="imu0" value="/cf1/imu"/>

    <!-- settings for using Twist -->
    <rosparam param="odom0_config">[false, false, false,
                                    false, false, false,
                                    true, true, true,
                                    false, false, false,
                                    false, false, false]</rosparam>

    <!-- <rosparam param="odom1_config">[false, false, true,
                                    false, false, false,
                                    false, false, false,
                                    false, false, false,
                                    false, false, false]</rosparam> -->

     <!-- settings for IMU -->
     <rosparam param="imu0_config">[false, false, false,
                                    true,  true,  true,
                                    false, false, false,
                                    true,  true,  true,
                                    true, true, true]</rosparam>

    <rosparam param="initial_state">[0.0, 0.0,  0.5,
                                    0.0,  0.0,  0.0,
                                    0.0,  0.0,  0.0,
                                    0.0,  0.0,  0.0,
                                    0.0,  0.0,  0.0]</rosparam>

      <param name="odom0_differential" value="false"/>
      <!-- <param name="odom1_differential" value="true"/> -->
      <param name="imu0_differential" value="false"/>
      <param name="imu0_remove_gravitational_acceleration" value="true"/>                                     

    <rosparam param="process_noise_covariance">[0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.06, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.03, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.03, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.06, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.025, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.0, 0.025, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.0, 0.0, 0.04, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.0, 0.0, 0.0, 0.01, 0.0, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01, 0.0, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02, 0.0, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01, 0.0, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01, 0.0,
                                                0.0, 0.0, 0.0, 0.0, 0.00, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.015]</rosparam>
    <remap from="/odometry/filtered" to="/odometry/visual" />
  </node>
</launch>